File: src/nylon/nylonobjects/NylonStack.java
package nylon.nylonobjects;

import nylon.NylonObject;

import java.util.Iterator;
import java.util.Stack;

/**
 * Created by Aedan Smith.
 */

public class NylonStack extends Stack<NylonObject> implements NylonObject {
    public NylonStack() {
    }

    public NylonStack(NylonObject nylonObject) {
        this.push(nylonObject);
    }

    @Override
    public boolean toBoolean(NylonStack stack) {
        return peek().toBoolean(stack);
    }

    @Override
    public double toDouble(NylonStack stack) {
        return peek().toDouble(stack);
    }

    @Override
    public Iterator<NylonObject> toIterator(NylonStack stack) {
        return iterator();
    }

    @Override
    public NylonStack toStack(NylonStack stack) {
        return this;
    }

    @Override
    public synchronized String toString() {
        for (NylonObject nylonObject : this) {
            if (nylonObject.getClass() != NylonCharacter.class) {
                return super.toString();
            }
        }
        String s = "";
        for (NylonObject nylonObject : this) {
            s += nylonObject.toString();
        }
        return s;
    }
}

File: src/nylon/NylonStack.kt
package nylon

import java.util.*

/**
 * Created by Aedan Smith.
 */

class NylonStack {
    val stacks = Stack<Stack<NylonObject<*>>>()

    init {
        stacks.add(Stack<NylonObject<*>>())
    }

    fun getStack() = try {
        stacks.peek()!!
    } catch (e: Exception) {
        throw NylonException("No active stack found")
    }

    fun peek(i: Int = 1) = try {
        getStack()[getStack().size - i]!!
    } catch (e: Exception) {
        throw NylonException("Could not find object at depth $i in stack ${getStack()}")
    }

    fun push(o: NylonObject<*>) = try {
        getStack().push(o)!!
    } catch (e: Exception) {
        throw NylonException("Could not push object $o to stack ${getStack()}")
    }

    fun pop() = try {
        getStack().pop()!!
    } catch (e: Exception) {
        throw NylonException("Could not pop object from stack ${getStack()}")
    }

    fun add(o: NylonObject<*>) = try {
        getStack().add(o)
    } catch (e: Exception) {
        throw NylonException("Could not add object $o to stack ${getStack()}")
    }

    fun add(i: Int, o: NylonObject<*>) = try {
        getStack().add(i, o)
    } catch (e: Exception) {
        throw NylonException("Could not add object $o to stack ${getStack()} at index $i")
    }

    fun removeAt(i: Int) = try {
        getStack().removeAt(i)!!
    } catch (e: Exception) {
        throw NylonException("Could not remove object at index $i from stack ${getStack()}")
    }

    fun getSize() = getStack().size

    fun iterator() = getStack().iterator()

    fun toList() = getStack().toList()

    fun clear() = getStack().clear()

    override fun toString() = getStack().toString()
}
